concordium-consensus/src/Concordium/KonsensusV1/Types.hs

```diff
@@ -0,0 +1,804 @@
+    blockParent :: d -> BlockHash
+    blockParent = qcBlock . blockQuorumCertificate
+
+    -- |'BakerId' of the baker of the block.
+    blockBaker :: d -> BakerId
+
+    -- |Signature verification key of the baker of the block.
+    blockBakerKey :: d -> BakerSignVerifyKey
+
+    -- |If the previous round timed-out, the timeout certificate for that round.
+    blockTimeoutCertificate :: d -> Optionally TimeoutCertificate
+
+    -- |If this block begins a new epoch, this is the finalization entry that finalizes the
+    -- trigger block.
+    blockEpochFinalizationEntry :: d -> Optionally FinalizationEntry
+
+    -- |The 'BlockNonce' generated by the baker's VRF.
+    blockNonce :: d -> BlockNonce
+
+    -- |The baker's signature on the block.
+    blockSignature :: d -> BlockSignature
+
+instance BakedBlockData Void where
+    blockQuorumCertificate = absurd
+    blockBaker = absurd
+    blockBakerKey = absurd
+    blockTimeoutCertificate = absurd
+    blockEpochFinalizationEntry = absurd
+    blockNonce = absurd
+    blockSignature = absurd
+
+-- |Projections for the data associated with a block (including a genesis block).
+class BlockData b where
+    -- |An associated type that should be an instance of 'BakedBlockData'.
+    -- This is returned by 'blockBakedData' for non-genesis blocks.
+    type BakedBlockDataType b
+
+    -- |Round number of the block.
+    blockRound :: b -> Round
+
+    -- |Epoch number of the block.
+    blockEpoch :: b -> Epoch
+
+    -- |Timestamp of the block.
+    blockTimestamp :: b -> Timestamp
+
+    -- |If the block is a baked (i.e. non-genesis) block, this returns the baked block data.
+    -- If the block is a genesis block, this returns 'Absent'.
+    blockBakedData :: b -> Optionally (BakedBlockDataType b)
+
+    -- |The list of transactions in the block.
+    blockTransactions :: b -> [BlockItem]
+
+    -- |The hash of the block state after executing the block.
+    blockStateHash :: b -> StateHash
+
+-- |A 'BakedBlock' consists of a non-genesis block, excluding the block signature.
+data BakedBlock = BakedBlock
+    { -- |Block round number.
+      bbRound :: !Round,
+      -- |Block epoch number.
+      bbEpoch :: !Epoch,
+      -- |Block nominal timestamp.
+      bbTimestamp :: !Timestamp,
+      -- |Block baker identity.
+      bbBaker :: !BakerId,
+      -- |Block baker signature verification key.
+      bbBakerKey :: !BakerSignVerifyKey,
+      -- |Quorum certificate of parent block.
+      bbQuorumCertificate :: !QuorumCertificate,
+      -- |Timeout certificate if the previous round timed-out.
+      bbTimeoutCertificate :: !(Optionally TimeoutCertificate),
+      -- |Epoch finalization entry if this is the first block in a new epoch.
+      bbEpochFinalizationEntry :: !(Optionally FinalizationEntry),
+      -- |Block nonce generated from the baker's VRF.
+      bbNonce :: !BlockNonce,
+      -- |Transactions in the block.
+      bbTransactions :: !(Vector.Vector BlockItem),
```

Comment:
Why is using a Vector here a good choice? The `blocktransactions` method in turn returns a list.